### ISSUE TITLE
Evaluating result based on the Top N words

### ISSUE THREAD

**Issue by TAsUjxnMIL** [MaartenGr/BERTopic#413]
Hello Maarten,
I have a corpus with 1000+ documents. Each document is labeled with its main topics. What I want to do is to calculate precision and recall. The problem I have is that the result is a set of words and I cannot get the generic term of the topics in order to compare to the labeling. 
An idea I had is to determine the average embedding of the top N words I get for a topic and compare this with the embedding of the labeling. I do not know if this is the right approach to calculate precision and recall or in general to check if the outcome fits with the label.

Did you provide any possibility to compare the outcome of BERTopic with the labeling? 

Thank you for your help!

**Comment by MaartenGr** [comment]
There are a number of things that make this question quite difficult to execute in practice. Comparing the outcome of BERTopic with labeling will typically only work in a number of cases. You can compare the generated clusters with RAND-based scores to see how it performs compared to the ground labels but that ignores the semantic content of the topics. 

The idea that you are proposing might work but do realize that the top N words are also weighted and taking the average might put too much emphasis on the tenth word compared to the first word. Also, you run the risk of overfitting the embedding of the labeling as it only describes a single word. If the labeling, however, does contain a sentence or description of the label then you get a much more accurate embedding of the label to compare to. Moreover, what will you do when you have 100 topics generated and there are 10 labels? Comparing cosine similarity could solve this issue but then you have to decide the minimum score. 

One trick that might be interesting to use is to run `.transform()` on the labels if you have descriptions for those labels. Topics will get assigned to the labels, and if you have set `calculate_probabilities=True` then you will also have the probability of each topic belonging to each label. Now that I think about it, to me, this seems to be the most accurate form of trying to match topics with labels. The most important thing here though is that you can describe each label in roughly two sentences. This is necessary to generate sufficiently accurate embeddings. 

Thus, although this might work in practice, there are a number of considerations that could greatly influence the quality of comparisons that you should take into account. Make sure to spend some time validating the mapping between topics and labels.

**Comment by TAsUjxnMIL** [comment]
Okay thank you Maarten I understand why my approach has some drawbacks. 
You approach seems quite accurate but what to do if the labels are just generic terms of the topics from each document. There is no description or sth. like that as a label. 
Do you think this approach would still be accurate? 

After performing .transform I have topics assigned to the labels. How to compare this topics from the labels with the top N words from the BERTopic result?

**Comment by MaartenGr** [comment]
Indeed, if you just have generic terms the resulting accuracy is likely to be lower. You can try it out with just the terms in hopes of finding accurate mappings but you will have to check yourself. 

First, you can use `.fit()` or `.fit_transform()` on the documents to generate the topics and then use `.transform()` on the labels to see which topics they typically belong to. 

If you do not have more extensive descriptions of your labels I would advise checking manually if the labels match the topics they are frequently assigned to and see if they make sense to you. Since you are checking the global topics and not each document it is likely you will only have to visually check tens of labels.

**Comment by TAsUjxnMIL** [comment]
Okay, but I don't think it is necessary to use .transform() on the labels because they are already the topics. 
So my question is if there is a procedure to check if the top N words fit to the label of a document. 

For example:
1.) BERTopic gives me the following for the first topic: 
     0:['football', 'sportsmen', 'hockey', 'nba', ...]

2.) The labeling of the first document is:
     ['sports']

Now I as a human can see that this topic creation was successful. But what I want is that my program determines that and says okay the top N words and the labeling fit together. So the model did the correct job for document one. 

Is this kind of procedure possible in code?

**Comment by MaartenGr** [comment]
Here, it might be interesting to focus more on word embedding techniques to do the cosine similarity, such as FastText since you are comparing words with each other. Using FastText, you can get a weighted embedding from the words in a BERTopic topic (weighted by cTFIDF value) which you can compare with the word embedding of the labeling. There is a strong assumption though that the labeling of the documents is clear enough to generate meaningful embeddings. 

Having said that, it then depends on what you are going to do with the resulting similarity score. What value are you going to use to indicate that a topic is related to a label? These values are up to you and requires some assumptions on your side.

**Comment by TAsUjxnMIL** [comment]
How to use FastText for weighted embedding? I cannot find something in the documentation about this kind of procedure.

What do you mean by
"There is a strong assumption though that the labeling of the documents is clear enough to generate meaningful embeddings." Does clear mean that it is not just one word?

**Comment by MaartenGr** [comment]
After you retrieve the FastText embeddings for each word, you can then average them. However, the c-TF-IDF score for one word is likely to be higher than another. Thus, instead of averaging them as if they were equally important, you can take the weighted average of these embeddings based on their c-TF-IDF value. The weighted average documentation of numpy can be found [here](https://numpy.org/doc/stable/reference/generated/numpy.average.html). 

Well, if you have labels for your documents and the labels give very little information, then it is likely that the resulting embedding is also not that accurate. For example, if you have a set of documents that are primarily about soccer and the resulting label is "sports", then the resulting word embedding of "sports" might not reflect the documents well. So you have to be sure, based on manual inspection, that the words used for the labels accurately describe the class the documents belong to.

**Comment by MaartenGr** [comment]
Due to inactivity, I'll be closing this issue. If you want to further discuss this issue, let me know and I'll make sure to re-open it.
